src/domain/core/src/services/dataset_registry.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::sync::Arc;
+
+use internal_error::InternalError;
+use opendatafabric::{AccountName, DatasetHandle, DatasetID, DatasetRef};
+use thiserror::Error;
+
+use crate::{Dataset, DatasetHandleStream, GetDatasetError, ResolvedDataset};
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait DatasetRegistry: Send + Sync {
+    fn all_dataset_handles(&self) -> DatasetHandleStream<'_>;
+
+    fn all_dataset_handles_by_owner(&self, owner_name: &AccountName) -> DatasetHandleStream<'_>;
+
+    async fn resolve_dataset_handle_by_ref(
+        &self,
+        dataset_ref: &DatasetRef,
+    ) -> Result<DatasetHandle, GetDatasetError>;
+
+    async fn resolve_multiple_dataset_handles_by_ids(
+        &self,
+        dataset_ids: Vec<DatasetID>,
+    ) -> Result<DatasetHandlesResolution, GetMultipleDatasetsError>;
+
+    fn get_dataset_by_handle(&self, dataset_handle: &DatasetHandle) -> Arc<dyn Dataset>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Extensions
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait DatasetRegistryExt: DatasetRegistry {
+    async fn try_resolve_dataset_handle_by_ref(
+        &self,
+        dataset_ref: &DatasetRef,
+    ) -> Result<Option<DatasetHandle>, InternalError>;
+
+    async fn get_dataset_by_ref(
+        &self,
+        dataset_ref: &DatasetRef,
+    ) -> Result<Arc<dyn Dataset>, GetDatasetError>;
+
+    fn get_resolved_dataset_by_handle(&self, dataset_handle: &DatasetHandle) -> ResolvedDataset;
```

Comment:
IMO it would simplify things a lot if we always returned `ResolvedDataset` everywhere - it would eliminate 2-3 methods from these interfaces.

Especially as per my other comment we make `ResolvedDataset` usable directly as a stand-in for `dyn Dataset` and allow simple conversion via `.into()`

Reply:
Done